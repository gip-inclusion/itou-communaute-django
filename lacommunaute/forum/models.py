from django.conf import settings
from django.contrib.contenttypes.fields import GenericRelation
from django.db import models
from django.urls import reverse
from django.utils.functional import cached_property
from machina.apps.forum.abstract_models import AbstractForum
from machina.models import DatedModel
from storages.backends.s3boto3 import S3Boto3Storage
from taggit.managers import TaggableManager

from lacommunaute.forum.enums import Kind as Forum_Kind
from lacommunaute.forum_conversation.models import Topic
from lacommunaute.forum_upvote.models import UpVote
from lacommunaute.partner.models import Partner
from lacommunaute.utils.validators import validate_image_size


class Forum(AbstractForum):
    kind = models.CharField(
        max_length=20, choices=Forum_Kind.choices, default=Forum_Kind.PUBLIC_FORUM, verbose_name="Type"
    )
    short_description = models.CharField(
        max_length=400, blank=True, null=True, verbose_name="Description courte (SEO)"
    )
    image = models.ImageField(
        storage=S3Boto3Storage(bucket_name=settings.AWS_STORAGE_BUCKET_NAME, file_overwrite=False),
        validators=[validate_image_size],
    )
    certified = models.BooleanField(default=False, verbose_name="Certifié par la communauté de l'inclusion")

    upvotes = GenericRelation(UpVote, related_query_name="forum")

    tags = TaggableManager()
    partner = models.ForeignKey(Partner, on_delete=models.CASCADE, null=True, blank=True)

    objects = models.Manager()

    def get_absolute_url(self):
        return reverse(
            "forum_extension:forum",
            kwargs={
                "pk": self.pk,
                "slug": self.slug,
            },
        )

    def get_unanswered_topics(self):
        return Topic.objects.unanswered().filter(forum__in=self.get_descendants(include_self=True))

    @cached_property
    def count_unanswered_topics(self):
        return self.get_unanswered_topics().count()

    def upvotes_count(self):
        return self.upvotes.count()

    @cached_property
    def is_in_documentation_area(self):
        return (self.type == Forum.FORUM_CAT and self.get_level() == 0) or (
            self.get_level() > 0 and self.get_ancestors().first().type == Forum.FORUM_CAT
        )

    @cached_property
    def is_toplevel_discussion_area(self):
        return self == Forum.objects.filter(lft=1, level=0).first()

    def get_session_rating(self, session_key):
        return getattr(ForumRating.objects.filter(forum=self, session_id=session_key).first(), "rating", None)

    def get_average_rating(self):
        return ForumRating.objects.filter(forum=self).aggregate(models.Avg("rating"))["rating__avg"]


class ForumRating(DatedModel):
    session_id = models.CharField(max_length=40)
    forum = models.ForeignKey(Forum, on_delete=models.CASCADE)
    rating = models.PositiveSmallIntegerField()
    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, null=True, blank=True)

    class Meta:
        verbose_name = "Notation Forum"
        verbose_name_plural = "Notations Forum"
        ordering = ("-created",)
