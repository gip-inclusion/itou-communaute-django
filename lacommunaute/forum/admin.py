from django.contrib import admin
from machina.apps.forum.admin import ForumAdmin as BaseForumAdmin

from lacommunaute.forum.models import ForumRating


class ForumAdmin(BaseForumAdmin):
    fieldsets = BaseForumAdmin.fieldsets
    fieldsets[0][1]["fields"] += ("short_description",)
    fieldsets[1][1]["fields"] += (
        "members_group",
        "invitation_token",
        "kind",
    )


@admin.register(ForumRating)
class ForumRatingAdmin(admin.ModelAdmin):
    list_display = ("forum", "rating", "created")
    list_filter = ("forum",)
    list_display_links = ("rating",)
    raw_id_fields = ("forum", "user")
