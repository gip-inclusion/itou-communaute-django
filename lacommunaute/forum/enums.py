from django.db import models


class Kind(models.TextChoices):
    PUBLIC_FORUM = "PUBLIC_FORUM", "Espace public"
    NEWS = "NEWS", "Actualit√©s"
