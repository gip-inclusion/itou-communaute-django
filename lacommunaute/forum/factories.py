import factory
from machina.core.db.models import get_model
from machina.test.factories.forum import ForumFactory as BaseForumFactory

from lacommunaute.forum.models import Forum
from lacommunaute.users.factories import GroupFactory


ForumPermission = get_model("forum_permission", "ForumPermission")
UserForumPermission = get_model("forum_permission", "UserForumPermission")


class ForumFactory(BaseForumFactory):
    type = Forum.FORUM_POST
    members_group = factory.SubFactory(GroupFactory, name=factory.SelfAttribute("..name"))
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=10)

    @factory.post_generation
    def with_public_perms(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        perms = [
            "can_see_forum",
            "can_read_forum",
            "can_start_new_topics",
            "can_reply_to_topics",
            "can_edit_own_posts",
            "can_delete_own_posts",
            "can_post_without_approval",
        ]

        anonymous_authorized_perms = [
            UserForumPermission(
                anonymous_user=True, authenticated_user=False, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]

        authentified_authorized_perms = [
            UserForumPermission(
                anonymous_user=False, authenticated_user=True, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]
        UserForumPermission.objects.bulk_create(anonymous_authorized_perms + authentified_authorized_perms)
