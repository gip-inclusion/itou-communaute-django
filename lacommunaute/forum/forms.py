import re

from django import forms
from django.conf import settings

from lacommunaute.forum.models import Forum


def wrap_iframe_in_div_tag(text):
    # iframe tags must be wrapped in a div tag to be displayed correctly
    # add div tag if not present

    iframe_regex = r"((<div>)?<iframe.*?</iframe>(</div>)?)"

    for match, starts_with, ends_with in re.findall(iframe_regex, text, re.DOTALL):
        if not starts_with and not ends_with:
            text = text.replace(match, f"<div>{match}</div>")

    return text


class ForumForm(forms.ModelForm):
    name = forms.CharField(required=True, label="Titre")
    short_description = forms.CharField(
        widget=forms.Textarea(attrs={"rows": 3}),
        max_length=400,
        required=True,
        label="Sous-titre (400 caractères pour le SEO)",
    )
    description = forms.CharField(
        widget=forms.Textarea(attrs={"rows": 20}), required=False, label="Contenu (markdown autorisé)"
    )
    image = forms.ImageField(
        required=False,
        label="Banniere de couverture, format 1200 x 630 pixels recommandé",
        widget=forms.FileInput(attrs={"accept": settings.SUPPORTED_IMAGE_FILE_TYPES.keys()}),
    )

    def save(self, commit=True):
        forum = super().save(commit=False)
        forum.description = wrap_iframe_in_div_tag(self.cleaned_data.get("description"))
        if commit:
            forum.save()
        return forum

    class Meta:
        model = Forum
        fields = ["name", "short_description", "description", "image"]
