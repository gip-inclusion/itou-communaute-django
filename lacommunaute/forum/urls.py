from django.urls import path
from machina.apps.forum.views import IndexView

from lacommunaute.forum.views import CategoryForumListView, ForumView


app_name = "forum_extension"


urlpatterns = [
    path("forum/<str:slug>-<int:pk>/", ForumView.as_view(), name="forum"),
    path("forum/documentation/", CategoryForumListView.as_view(), name="documentation"),
    path("forum/", IndexView.as_view(), name="index"),
]
