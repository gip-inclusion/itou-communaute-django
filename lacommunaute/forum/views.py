import logging

from django.conf import settings
from django.contrib.auth.mixins import UserPassesTestMixin
from django.contrib.contenttypes.models import ContentType
from django.db.models.query import QuerySet
from django.shortcuts import get_object_or_404, render
from django.urls import reverse, reverse_lazy
from django.views import View
from django.views.generic import CreateView, ListView, UpdateView
from machina.apps.forum.views import ForumView as BaseForumView
from machina.core.loading import get_class
from taggit.models import Tag

from lacommunaute.forum.forms import ForumForm
from lacommunaute.forum.models import Forum, ForumRating
from lacommunaute.forum_conversation.forms import PostForm
from lacommunaute.forum_conversation.view_mixins import FilteredTopicsListViewMixin
from lacommunaute.forum_upvote.models import UpVote
from lacommunaute.utils.perms import add_public_perms_on_forum, forum_visibility_content_tree_from_forums


logger = logging.getLogger(__name__)

PermissionRequiredMixin = get_class("forum_permission.viewmixins", "PermissionRequiredMixin")


class SubCategoryForumListMixin:
    def get_descendants(self):
        qs = self.get_forum().get_descendants()

        forum_tag = self.request.GET.get("forum_tag") or None
        if forum_tag:
            qs = qs.filter(tags__slug=forum_tag)

        return qs.prefetch_related("tags")

    def get_tags_of_descendants(self):
        return Tag.objects.filter(
            taggit_taggeditem_items__content_type=ContentType.objects.get_for_model(Forum),
            taggit_taggeditem_items__object_id__in=self.get_forum().get_descendants().values_list("id", flat=True),
        ).distinct()

    def forum_tag_context(self):
        return {
            # TODO : remove permission management, though all forums are public in our case
            "sub_forums": forum_visibility_content_tree_from_forums(self.request, self.get_descendants()),
            "tags_of_descendants": self.get_tags_of_descendants(),
            "active_forum_tag_slug": self.request.GET.get("forum_tag") or None,
        }


class ForumView(BaseForumView, FilteredTopicsListViewMixin, SubCategoryForumListMixin):
    paginate_by = settings.FORUM_TOPICS_NUMBER_PER_PAGE

    def get_template_names(self):
        if self.request.META.get("HTTP_HX_REQUEST"):
            return ["forum_conversation/topic_list.html"]
        if self.will_render_documentation_variant():
            return ["forum/forum_documentation.html"]
        if self.will_render_documentation_category_variant():
            return ["forum/forum_documentation_category.html"]
        return ["forum/forum_detail.html"]

    def will_render_documentation_variant(self):
        return self.get_forum().parent and self.forum.is_in_documentation_area

    def will_render_documentation_category_variant(self):
        return self.get_forum().is_in_documentation_area and self.forum.level == 0

    def get_queryset(self):
        return self.filter_queryset(self.get_forum().topics.optimized_for_topics_list(self.request.user.id))

    def get_context_data(self, **kwargs):
        forum = self.get_forum()

        if self.request.user.is_authenticated:
            forum.has_upvoted = UpVote.objects.filter(
                object_id=forum.id,
                voter=self.request.user,
                content_type_id=ContentType.objects.get_for_model(forum).id,
            ).exists()

        context = super().get_context_data(**kwargs)
        context["forum"] = forum
        context["FORUM_NUMBER_POSTS_PER_TOPIC"] = settings.FORUM_NUMBER_POSTS_PER_TOPIC
        context["next_url"] = reverse("forum_extension:forum", kwargs={"pk": forum.pk, "slug": self.forum.slug})
        context["loadmoretopic_url"] = self.get_load_more_url(
            reverse("forum_extension:forum", kwargs={"pk": forum.pk, "slug": self.forum.slug})
        )
        context["loadmoretopic_suffix"] = "topicsinforum"
        context["form"] = PostForm(forum=forum, user=self.request.user)

        context["rating"] = forum.get_session_rating(self.request.session.session_key)

        context["filter_dropdown_endpoint"] = (
            None
            if self.request.GET.get("page")
            else reverse("forum_extension:forum", kwargs={"pk": forum.pk, "slug": self.forum.slug})
        )
        context = context | self.get_topic_filter_context()

        if self.will_render_documentation_category_variant():
            context = context | self.forum_tag_context()

        if self.will_render_documentation_variant():
            context["sibling_forums"] = forum.get_siblings(include_self=True)

        if forum.image:
            context["og_image"] = forum.image
        return context


class SubCategoryForumListView(BaseForumView, SubCategoryForumListMixin):
    template_name = "forum/partials/subcategory_forum_list.html"

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs) | self.forum_tag_context()
        return context


class ForumUpdateView(UserPassesTestMixin, UpdateView):
    template_name = "forum/forum_create_or_update.html"
    form_class = ForumForm
    model = Forum

    def test_func(self):
        return self.request.user.is_staff

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["title"] = f"Mettre à jour le forum {self.object.name}"
        context["back_url"] = reverse("forum_extension:forum", kwargs={"pk": self.object.pk, "slug": self.object.slug})
        return context


class CategoryForumListView(ListView):
    template_name = "forum/category_forum_list.html"
    context_object_name = "forums"

    def get_queryset(self) -> QuerySet[Forum]:
        return Forum.objects.filter(type=Forum.FORUM_CAT, level=0)


class BaseCategoryForumCreateView(UserPassesTestMixin, CreateView):
    template_name = "forum/forum_create_or_update.html"
    form_class = ForumForm

    def test_func(self):
        return self.request.user.is_staff

    def form_valid(self, form):
        response = super().form_valid(form)
        add_public_perms_on_forum(form.instance)
        return response


class CategoryForumCreateView(BaseCategoryForumCreateView):
    success_url = reverse_lazy("forum_extension:documentation")

    def form_valid(self, form):
        form.instance.parent = None
        form.instance.type = Forum.FORUM_CAT
        return super().form_valid(form)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["title"] = "Créer une nouvelle catégorie documentaire"
        context["back_url"] = reverse("forum_extension:documentation")
        return context


class SubCategoryForumCreateView(BaseCategoryForumCreateView):
    def get_success_url(self):
        return reverse("forum_extension:forum", kwargs={"pk": self.object.pk, "slug": self.object.slug})

    def get_parent_forum(self):
        return Forum.objects.get(pk=self.kwargs["pk"])

    def form_valid(self, form):
        form.instance.type = Forum.FORUM_POST
        form.instance.parent = self.get_parent_forum()
        return super().form_valid(form)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["title"] = f"Créer une fiche pratique dans la catégorie {self.get_parent_forum().name}"
        context["back_url"] = reverse(
            "forum_extension:forum", kwargs={"pk": self.get_parent_forum().pk, "slug": self.get_parent_forum().slug}
        )
        return context


class ForumRatingView(View):
    def post(self, request, *args, **kwargs):
        forum_rating = ForumRating.objects.create(
            forum=get_object_or_404(Forum, pk=self.kwargs["pk"]),
            user=request.user if request.user.is_authenticated else None,
            rating=int(request.POST["rating"]),
            session_id=request.session.session_key,
        )

        return render(
            request,
            "forum/partials/rating.html",
            context={"forum": forum_rating.forum, "rating": forum_rating.rating},
        )
