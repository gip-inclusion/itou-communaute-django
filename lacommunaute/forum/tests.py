from dateutil.relativedelta import relativedelta
from django.contrib.auth.models import AnonymousUser
from django.db import IntegrityError
from django.test import RequestFactory, TestCase
from django.urls import reverse
from django.utils import timezone
from django.utils.http import urlencode
from faker import Faker
from machina.core.db.models import get_model
from machina.core.loading import get_class

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.forum.views import ForumView
from lacommunaute.forum_conversation.factories import PostFactory, TopicFactory
from lacommunaute.forum_conversation.forms import PostForm
from lacommunaute.forum_conversation.forum_polls.factories import TopicPollVoteFactory
from lacommunaute.forum_conversation.models import Topic
from lacommunaute.forum_upvote.factories import UpVoteFactory
from lacommunaute.users.factories import UserFactory


faker = Faker()

ForumPermission = get_model("forum_permission", "ForumPermission")
UserForumPermission = get_model("forum_permission", "UserForumPermission")
PermissionHandler = get_class("forum_permission.handler", "PermissionHandler")
assign_perm = get_class("forum_permission.shortcuts", "assign_perm")
remove_perm = get_class("forum_permission.shortcuts", "remove_perm")


class ForumViewQuerysetTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.user = UserFactory()
        cls.forum = ForumFactory()
        cls.view = ForumView()
        cls.view.kwargs = {"pk": cls.forum.pk}
        cls.view.request = RequestFactory().get("/")
        cls.view.request.user = cls.user

    def test_excluded_announces_topics(self):
        TopicFactory(forum=self.forum, poster=self.user, type=Topic.TOPIC_ANNOUNCE)
        self.assertFalse(self.view.get_queryset())

    def test_exclude_not_approved_posts(self):
        TopicFactory(forum=self.forum, poster=self.user, approved=False)
        self.assertFalse(self.view.get_queryset())

    def test_ordering_topics_on_last_post(self):
        old_topic = TopicFactory(forum=self.forum, poster=self.user)
        new_topic = TopicFactory(forum=self.forum, poster=self.user)

        PostFactory(topic=old_topic, poster=self.user)
        PostFactory(topic=new_topic, poster=self.user)

        qs = self.view.get_queryset()
        self.assertEqual(qs.first(), new_topic)
        self.assertEqual(qs.last(), old_topic)

        PostFactory(topic=old_topic, poster=self.user)
        self.assertEqual(qs.first(), old_topic)
        self.assertEqual(qs.last(), new_topic)

    def test_pagination(self):
        self.assertEqual(10, self.view.paginate_by)

    def test_has_liked(self):
        TopicFactory(forum=self.forum, poster=self.user, with_like=True)

        first_topic = self.view.get_queryset().first()
        self.assertEqual(first_topic.likes, 1)
        self.assertTrue(first_topic.has_liked)

    def test_has_not_liked(self):
        TopicFactory(forum=self.forum, poster=self.user)

        first_topic = self.view.get_queryset().first()
        self.assertEqual(first_topic.likes, 0)
        self.assertFalse(first_topic.has_liked)


class ForumViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.topic = TopicFactory(with_post=True)
        cls.user = cls.topic.poster
        cls.forum = cls.topic.forum

        # Assign some permissions
        cls.perm_handler = PermissionHandler()
        assign_perm("can_read_forum", cls.user, cls.forum)
        assign_perm("can_see_forum", cls.user, cls.forum)
        assign_perm("can_post_without_approval", cls.user, cls.forum)
        assign_perm("can_reply_to_topics", cls.user, cls.forum)

        cls.url = reverse("forum:forum", kwargs={"pk": cls.forum.pk, "slug": cls.forum.slug})

    def test_subscription_button_is_hidden(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(
            response,
            reverse(
                "forum_conversation_extension:post_create",
                kwargs={
                    "forum_pk": self.forum.pk,
                    "forum_slug": self.forum.slug,
                    "pk": self.topic.pk,
                    "slug": self.topic.slug,
                },
            ),
        )
        self.assertNotContains(
            response,
            reverse(
                "forum_member:topic_subscribe",
                kwargs={
                    "pk": self.topic.pk,
                },
            ),
        )

    def test_show_comments(self):
        topic_url = reverse(
            "forum_conversation_extension:showmore_posts",
            kwargs={
                "forum_pk": self.forum.pk,
                "forum_slug": self.forum.slug,
                "pk": self.topic.pk,
                "slug": self.topic.slug,
            },
        )
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertNotContains(response, f'<a href="{topic_url}"')

        PostFactory.create_batch(2, topic=self.topic, poster=self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(response, f'hx-get="{topic_url}"')
        self.assertContains(response, "Voir les 2 autres rÃ©ponses")

    def test_show_more_content(self):
        topic = TopicFactory(
            with_post=True, poster=self.user, forum=self.forum, post__content=faker.paragraph(nb_sentences=100)
        )
        topic_url = reverse(
            "forum_conversation_extension:showmore_topic",
            kwargs={
                "forum_pk": topic.forum.pk,
                "forum_slug": topic.forum.slug,
                "pk": topic.pk,
                "slug": topic.slug,
            },
        )
        self.client.force_login(self.user)
        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertContains(response, f'hx-get="{topic_url}"')
        self.assertContains(response, "+ voir la suite")

    def test_has_liked(self):
        TopicFactory(forum=self.forum, poster=self.user, with_post=True, with_like=True)

        self.client.force_login(self.user)
        response = self.client.get(self.url)
        # icon: solid heart
        self.assertContains(response, '<i class="ri-heart-3-fill" aria-hidden="true"></i>')
        self.assertContains(response, "<span>1 J'aime</span>")

    def test_has_not_liked(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        # icon: regular heart (outlined)
        self.assertContains(response, '<i class="ri-heart-3-line" aria-hidden="true"></i>')
        self.assertContains(response, "<span>0 J'aime</span>")

    def test_anonymous_like(self):
        assign_perm("can_read_forum", AnonymousUser(), self.topic.forum)
        params = {"next_url": self.url}
        url = f"{reverse('inclusion_connect:authorize')}?{urlencode(params)}"

        response = self.client.get(self.url)
        self.assertContains(response, url)

        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertNotContains(response, url)

    def test_moderator_links(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertNotContains(
            response,
            reverse(
                "forum_extension:engagement",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )
        self.assertNotContains(
            response,
            reverse(
                "members:forum_profiles",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )

        assign_perm("can_approve_posts", self.user, self.forum)
        response = self.client.get(self.url)
        self.assertContains(
            response,
            reverse(
                "forum_extension:engagement",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )
        self.assertContains(
            response,
            reverse(
                "members:forum_profiles",
                kwargs={"pk": self.forum.pk, "slug": self.forum.slug},
            ),
        )

    def test_poll_form(self):
        topic = TopicFactory(forum=self.forum, poster=self.user, with_post=True, with_poll_vote=True)
        poll_option = topic.poll.options.first()
        self.client.force_login(self.user)

        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)
        self.assertContains(response, poll_option.poll.question)
        self.assertContains(response, poll_option.text)

    def test_postform_in_context(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)
        self.assertIsInstance(response.context_data["form"], PostForm)
        self.assertContains(response, f'id="collapsePost{self.topic.pk}')

    def test_cannot_submit_post(self):
        user = UserFactory()
        assign_perm("can_read_forum", user, self.forum)
        assign_perm("can_see_forum", user, self.forum)
        remove_perm("can_reply_to_topics", user, self.forum)
        self.client.force_login(user)

        response = self.client.get(self.url)

        self.assertEqual(response.status_code, 200)
        self.assertNotContains(response, f'id="collapsePost{self.topic.pk}')


class ForumModelTest(TestCase):
    def test_invitation_token_is_unique(self):
        forum = ForumFactory()

        with self.assertRaises(IntegrityError):
            forum.id = None
            forum.save()

    def test_get_stats(self):

        forum = ForumFactory()
        topic = TopicFactory(forum=forum, with_post=True, with_poll_vote=True)
        UpVoteFactory(post=topic.posts.first(), voter=topic.poster)

        # create old instance to ensure they are filtered
        fifteen_days_ago = timezone.now() - relativedelta(days=15)
        old_topic = TopicFactory(forum=forum)
        old_topic.created = fifteen_days_ago
        old_topic.save()
        old_post = PostFactory(topic=old_topic, poster=old_topic.poster)
        old_post.created = fifteen_days_ago
        old_post.save()
        old_pollvote = TopicPollVoteFactory(voter=old_topic.poster, poll_option__poll__topic=old_topic)
        old_pollvote.timestamp = fifteen_days_ago
        old_pollvote.save()
        old_upvote = UpVoteFactory(post=old_topic.posts.first(), voter=old_topic.poster)
        old_upvote.created_at = fifteen_days_ago
        old_upvote.save()

        # create fake forum to ensure post, topic, etcâ¦ number is filter by forum
        fake_forum = ForumFactory()
        fake_topic = TopicFactory(forum=fake_forum, with_post=True, with_poll_vote=True)
        UpVoteFactory(post=fake_topic.posts.first(), voter=fake_topic.poster)

        self.assertEqual(
            forum.get_stats(1),
            {
                "period": [timezone.now().strftime("%Y-%W")],
                "pollvotes": [1],
                "upvotes": [1],
                "posts": [1],
                "topics": [1],
            },
        )

    def test_count_engaged_users(self):
        # first user posts, likes, votes
        topic = TopicFactory(with_post=True, with_like=True, with_poll_vote=True)

        # second user posts
        PostFactory(topic=topic, with_upvote=True)

        # anonymous user posts
        PostFactory(topic=topic, anonymous=True)

        # post, like, vote in an other forum = ignored in count
        TopicFactory(with_post=True, with_like=True, with_poll_vote=True)

        self.assertEqual(
            topic.forum.count_engaged_users,
            {
                "posters": 2,
                "likers": 1,
                "voters": 1,
                "upvoters": 1,
                "authenticated_users": 2,
                "anonymous_posters": 1,
                "all_users": 3,
            },
        )
