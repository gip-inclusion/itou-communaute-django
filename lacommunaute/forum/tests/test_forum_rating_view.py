import pytest
from django.urls import reverse

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.forum.models import ForumRating
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup


@pytest.fixture(name="forum")
def fixture_forum(db):
    return ForumFactory()


def test_get_forum_rating_view(client, db, forum):
    response = client.get(reverse("forum_extension:rate", kwargs={"pk": forum.pk, "slug": forum.slug}))
    assert response.status_code == 405


def test_anonymous_post_forum_rating_view(client, db, forum, snapshot):
    client.session.save()

    response = client.post(
        reverse("forum_extension:rate", kwargs={"pk": forum.pk, "slug": forum.slug}),
        data={"rating": "5"},
    )
    assert response.status_code == 200
    assert response.context["forum"] == forum
    assert response.context["rating"] == 5
    content = parse_response_to_soup(response, replace_in_href=[forum])
    assert str(content) == snapshot(name="anonymous_post_forum_rating_view")

    forum_rating = ForumRating.objects.get()
    assert forum_rating.forum == forum
    assert forum_rating.user is None
    assert forum_rating.rating == 5
    assert forum_rating.session_id == client.session.session_key


def test_authenticated_post_forum_rating_view(client, db, forum, snapshot):
    user = UserFactory()
    client.force_login(user)

    response = client.post(
        reverse("forum_extension:rate", kwargs={"pk": forum.pk, "slug": forum.slug}),
        data={"rating": "1"},
    )
    assert response.status_code == 200
    content = parse_response_to_soup(response, replace_in_href=[forum])
    assert str(content) == snapshot(name="authenticated_post_forum_rating_view")

    forum_rating = ForumRating.objects.get()
    assert forum_rating.forum == forum
    assert forum_rating.user == user
    assert forum_rating.rating == 1
    assert forum_rating.session_id == client.session.session_key
