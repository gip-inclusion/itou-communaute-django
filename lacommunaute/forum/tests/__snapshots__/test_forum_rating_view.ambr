# serializer version: 1
# name: test_anonymous_post_forum_rating_view[anonymous_post_forum_rating_view]
  '''
  <html><head></head><body><div id="rating-area">
      <h2 class="mt-3">Cette fiche vous a-t-elle été utile ?</h2>
      
          
              
                  ★
              
          
              
                  ★
              
          
              
                  ★
              
          
              
                  ★
              
          
              
                  ★
              
          
      
  </div>
  </body></html>
  '''
# ---
# name: test_authenticated_post_forum_rating_view[authenticated_post_forum_rating_view]
  '''
  <html><head></head><body><div id="rating-area">
      <h2 class="mt-3">Cette fiche vous a-t-elle été utile ?</h2>
      
          
              
                  ★
              
          
              
                  -
              
          
              
                  -
              
          
              
                  -
              
          
              
                  -
              
          
      
  </div>
  </body></html>
  '''
# ---
